sw/airborne/modules/nav/nav_rotorcraft_hybrid.c

```diff
@@ -171,21 +176,65 @@ static bool nav_hybrid_approaching(struct EnuCoor_f *wp, struct EnuCoor_f *from,
   return false;
 }
 
-#if 0 // TODO reuse existing patterns for now
-
-static void nav_circle(struct EnuCoor_f *wp_center, float radius)
+static void nav_hybrid_circle(struct EnuCoor_f *wp_center, float radius)
 {
+  struct FloatVect2 pos_diff;
+  float desired_speed;
 
-  // implement nav hybrid circle
+  VECT2_DIFF(pos_diff, *stateGetPositionEnu_f(), *wp_center);
+  // direction of rotation
+  float sign_radius = radius > 0.f ? 1.f : -1.f;
+  // absolute radius
+  float abs_radius = fabsf(radius);
+
+  if (abs_radius > 0.1f) {
+    // store last qdr
+    float last_qdr = nav_rotorcraft_base.circle.qdr;
+    // compute qdr
+    nav_rotorcraft_base.circle.qdr = atan2f(pos_diff.y, pos_diff.x);
+    // increment circle radians
+    float trigo_diff = nav_rotorcraft_base.circle.qdr - last_qdr;
+    NormRadAngle(trigo_diff);
+    nav_rotorcraft_base.circle.radians += trigo_diff;
+    // progress angle
+    float progess_angle = GUIDANCE_INDI_NAV_CIRCLE_DIST / abs_radius;
+    Bound(progess_angle, M_PI / 16.f, M_PI / 4.f);
+    float alpha = nav_rotorcraft_base.circle.qdr - sign_radius * progess_angle;
+    // final target position, should be on the circle, for display
+    nav.target.x = wp_center->x + cosf(alpha) * abs_radius;
+    nav.target.y = wp_center->y + sinf(alpha) * abs_radius;
+  }
+  else {
+    // radius is too small, direct to center
+    VECT2_COPY(nav.target, *wp_center);
+  }
+  // compute desired speed
+  if (force_forward) {
+    desired_speed = nav_max_speed;
+  } else {
+    float radius_diff = fabsf(float_vect2_norm(&pos_diff) - abs_radius);
+    if (radius_diff > GUIDANCE_INDI_NAV_CIRCLE_DIST) {
+      // far from circle, speed proportional to diff
+      desired_speed = radius_diff * gih_params.pos_gain;
+    } else {
+      // close to circle, speed function of radius for a feasible turn
+      // MAX_BANK / 2 gives some margins for the turns
+      desired_speed = sqrtf(PPRZ_ISA_GRAVITY * abs_radius * tanf(GUIDANCE_H_MAX_BANK / 2.f));
```

Comment:
Nice idea, but maybe some safety margin should be taken here?

Reply:
the max_bank divided by 2 is suppose to be the margin here

Reply:
Ah now I see, ok!